### ISSUE TITLE
Problem about weight initialization using DDP

### ISSUE THREAD

**Issue by K-Nick** [facebookresearch/detr#544]
It seems that you set a different seed for each rank before building the model. This may lead to different parameter initialization for different duplicate on each rank. Is it a mistake or a deliberate design?

Here is a comment from pytorch lightning ddp advice
> Setting all the random seeds to the same value. This is important in a distributed training setting. Each rank will get its own set of initial weights. If they don't match up, the gradients will not match either, leading to training that may not converge.

```python
"""starts from main.py line 347"""
# fix the seed for reproducibility
seed = args.seed + dist.get_rank()
torch.manual_seed(seed)
np.random.seed(seed)
random.seed(seed)
# torch.set_deterministic(True)
torch.use_deterministic_algorithms(True)

# Build the model
model, criterion, weight_dict = build_model(args)
model.to(device)
```

**Comment by xk-huang** [comment]
Check here. All the parameters are forced to be same when DDP instantiated. https://github.com/pytorch/pytorch/blob/1dba81f56dc33b44d7b0ecc92a039fe32ee80f8d/torch/nn/parallel/distributed.py#LL798C63-L798C63
